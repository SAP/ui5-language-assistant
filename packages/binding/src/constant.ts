export const BINDING_ISSUE_TYPE = "binding-issue";
export const PROPERTY_BINDING_INFO =
  "sap.ui.base.ManagedObject.PropertyBindingInfo";
export const AGGREGATION_BINDING_INFO =
  "sap.ui.base.ManagedObject.AggregationBindingInfo";
export const Binding_Mode = "sap.ui.model.BindingMode";
