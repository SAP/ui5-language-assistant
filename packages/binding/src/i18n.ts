import type { i18n, TFunction } from "i18next";
import { DEFAULT_I18N_NAMESPACE } from "@ui5-language-assistant/context";
import i18nEn from "./i18n/i18n.json";
import i18next from "i18next";

export function initI18n(i18nInstance: i18n): void {
  i18nInstance.addResourceBundle(
    i18nInstance.language,
    DEFAULT_I18N_NAMESPACE,
    i18nEn
  );
}

export const adaptTranslatedText = (text: string): string => {
  if (/&quot;/.test(text)) {
    text = text.replace(/&quot;/g, '"');
  }
  if (/&#39;/.test(text)) {
    text = text.replace(/&#39;/g, "'");
  }
  if (/&amp;quot;/.test(text)) {
    text = text.replace(/&amp;quot;/g, "&quot;");
  }
  return text;
};

export const t: TFunction = (key: string, ...args) => {
  const result = i18next.t(key, ...args);
  return adaptTranslatedText(result);
};
