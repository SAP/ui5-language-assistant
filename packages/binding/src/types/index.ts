export type { BindingIssue } from "./issues";
export type {
  PropertyBindingInfoElement,
  Dependents,
  PropertyType,
  BindContext,
} from "./binding-info";
export { BindingInfoName, TypeKind } from "./binding-info";

export { BINDING_ISSUE_TYPE } from "../../src/constant";

export type {
  CursorContext,
  KeyValueContext,
  BaseContext,
  EmptyContext,
  InitialContext,
  KeyContext,
  ValueContext,
} from "./cursor";
