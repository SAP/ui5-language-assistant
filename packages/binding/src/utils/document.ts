import type { OffsetRange } from "@ui5-language-assistant/logic-utils";
import type { Range } from "vscode-languageserver-types";

export const rangeToOffsetRange = (range?: Range): OffsetRange => {
  if (!range) {
    return { start: 0, end: 0 };
  }
  return { start: range.start.character, end: range.end.character };
};
