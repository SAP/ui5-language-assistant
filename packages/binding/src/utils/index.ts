export {
  typesToValue,
  valueTypeMap,
  isParts,
  findRange,
  defaultRange,
  getPropertyTypeWithPossibleValue,
} from "./element";

export { getCursorContext } from "./cursor";

export { getLogger } from "./logger";
