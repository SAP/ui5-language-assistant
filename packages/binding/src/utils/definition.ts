import {
  UI5Aggregation,
  UI5Class,
  UI5ConstructorParameters,
  UI5Type,
} from "@ui5-language-assistant/semantic-model-types";
import {
  BindContext,
  BindingInfoElement,
  BindingInfoName,
  ClassName,
  Dependents,
  PropertyType,
  TypeKind,
} from "../types";
import { FILTER_OPERATOR } from "../constant";
import { ui5NodeToFQN } from "@ui5-language-assistant/logic-utils";
import { getSorterPossibleElement } from "../definition/sorter";
import { getFiltersPossibleElement } from "../definition/filter";
import { getDocumentation } from "./documentation";
import { forOwn } from "lodash";

const notAllowedElements: Map<BindingInfoName, BindingInfoName[]> = new Map([
  [BindingInfoName.path, [BindingInfoName.parts, BindingInfoName.value]],
  [BindingInfoName.value, [BindingInfoName.parts, BindingInfoName.path]],
  [BindingInfoName.parts, [BindingInfoName.value, BindingInfoName.path]],
]);
const dependents: Map<BindingInfoName, Dependents[]> = new Map([
  [
    BindingInfoName.constraints,
    [
      {
        name: BindingInfoName.type,
        type: [
          {
            kind: TypeKind.string,
            dependents: [],
            notAllowedElements: [],
          },
        ],
      },
    ],
  ],
  [
    BindingInfoName.formatOptions,
    [
      {
        name: BindingInfoName.type,
        type: [
          {
            kind: TypeKind.string,
            dependents: [],
            notAllowedElements: [],
          },
        ],
      },
    ],
  ],
]);
const defaultBoolean: Map<string, boolean[]> = new Map([
  ["Boolean", [true, false]],
  ["boolean", [true, false]],
]);

const classKind: Map<ClassName, TypeKind> = new Map([
  [ClassName.Sorter, TypeKind.object],
  [ClassName.Filter, TypeKind.object],
]);

/**
 * Retrieves possible values from UI5 class which is extended.
 *
 * @param {BindContext} context - The binding context containing the UI5 model.
 * @param {UI5Class} type - The UI5 class type to find possible values for.
 * @returns {string[]} An array of possible values for the specified UI5 class.
 */
export const getPossibleValuesForClass = (
  context: BindContext,
  type: UI5Class
): string[] => {
  const result: string[] = [];
  forOwn(context.ui5Model.classes, (value, key) => {
    let clzExtends = value.extends;
    while (clzExtends !== undefined) {
      if (clzExtends.name === type.name && clzExtends.kind === type.kind) {
        result.push(key);
        break;
      }
      clzExtends = clzExtends.extends;
    }
  });

  return result;
};

/**
 * Currently [api.json](https://ui5.sap.com/1.118.1/test-resources/sap/ui/core/designtime/api.json) provides these constructor parameters as old school convention
 * e.g `sPath` for `path` where `s` stands for string type. These params are [map in runtime](https://github.com/SAP/openui5/blob/master/src/sap.ui.core/src/sap/ui/model/Sorter.js#L54-L60).
 * We build a map to overcome old school convention
 */
const sorterMap = new Map([
  ["sPath", "path"],
  ["path", "path"],
  ["bDescending", "descending"],
  ["descending", "descending"],
  ["vGroup", "group"],
  ["group", "group"],
  ["fnComparator", "comparator"],
  ["comparator", "comparator"],
]);

/**
 * Retrieves possible binding information elements for a given UI5 class.
 *
 * @param {Object} param - The parameters for retrieving possible elements.
 * @param {BindContext} param.context - The binding context containing the UI5 model.
 * @param {UI5Aggregation} [param.ui5Aggregation] - The UI5 aggregation information.
 * @param {boolean} [param.forHover=false] - Flag indicating if the elements are for hover information.
 * @param {UI5Class} param.type - The UI5 class type to find possible elements for.
 * @returns {BindingInfoElement[]} An array of possible binding information elements for the specified UI5 class.
 */
export const getPossibleElement = (param: {
  context: BindContext;
  ui5Aggregation?: UI5Aggregation;
  forHover?: boolean;
  type: UI5Class;
}): BindingInfoElement[] => {
  const result: BindingInfoElement[] = [];
  /* istanbul ignore next */
  const { ui5Aggregation, forHover = false, type, context } = param;
  if (type.name === ClassName.Sorter) {
    const parameters = type.ctor && type.ctor.parameters;
    if (!parameters) {
      return getSorterPossibleElement();
    }
    for (const constParam of parameters) {
      if (!constParam.type) {
        continue;
      }
      const name = sorterMap.get(constParam.name) ?? constParam.name;
      if (name === "vSorterInfo" && constParam.parameterProperties) {
        result.push(
          ...getConstructorParameterProperties({
            ...param,
            parameterProperties: constParam.parameterProperties,
          })
        );
        continue;
      }
      const reference = getReference(constParam.type);
      const data: BindingInfoElement = {
        name,
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        type: buildType({
          context,
          type: constParam.type,
          name,
          collection: false,
          ui5Aggregation,
          forHover,
          reference,
        }),
        documentation: getDocumentation({
          context,
          prop: constParam,
          FQN: ui5NodeToFQN(type),
          titlePrefix: "(class)",
          forHover,
        }),
      };
      data.required = !constParam.optional;
      result.push(data);
    }
  }

  if (type.name === ClassName.Filter) {
    // for filters only try `vFilterInfo` from constructor
    /* istanbul ignore next */
    const vFilter = type.ctor?.parameters?.find(
      (i) => i.name === "vFilterInfo"
    );
    if (!vFilter) {
      // use fallback filter
      return getFiltersPossibleElement();
    }
    result.push(
      ...getConstructorParameterProperties({
        ...param,
        parameterProperties: vFilter.parameterProperties,
      })
    );
  }
  return result;
};

/**
 * Retrieves class kind based on the provided class name.
 * If the class name is not found in the classKind map, it falls back to TypeKind.string.
 */
const getClassKind = (name: string) =>
  classKind.get(name as ClassName) ?? TypeKind.string;

/**
 * Retrieves a reference to other element based on the provided UI5Type. Currently only for `filters`.
 *
 * @param {UI5Type} type - The type object to get the reference for.
 * @returns {string | undefined} The reference string if found, otherwise undefined.
 */
export function getReference(type: UI5Type): string | undefined {
  let reference;
  switch (type.kind) {
    case "PrimitiveType":
    case "UI5Enum":
    case "UI5Class":
    case "UI5Typedef":
      if (type.name === ClassName.Filter) {
        reference = "filters";
      }
      break;
    case "ArrayType":
      if (type.type) {
        reference = getReference(type.type);
      }
      break;
    case "UnionType":
      for (const t of type.types) {
        reference = getReference(t);
        if (reference) {
          break;
        }
      }
      break;
    default:
      break;
  }
  return reference;
}

const getFromMap = <T, U extends string>(
  map: Map<U, T[]>,
  name: U,
  /* istanbul ignore next */
  aggregation = false
): T[] => {
  return aggregation ? [] : map.get(name) ?? [];
};

/**
 * Builds a PropertyType array based on the provided parameters.
 *
 * @param {Object} param - The parameters for building the PropertyType array.
 * @param {BindContext} param.context - The binding context.
 * @param {UI5Type} param.type - The UI5 type.
 * @param {string} param.name - The name of the property.
 * @param {boolean} [param.collection=false] - Indicates if the property is a collection.
 * @param {UI5Aggregation} [param.ui5Aggregation] - The UI5 aggregation.
 * @param {boolean} [param.forHover=false] - Indicates if the property is for hover.
 * @param {string} [param.reference] - The reference string.
 * @returns {PropertyType[]} The built PropertyType array.
 */
export const buildType = (param: {
  context: BindContext;
  type: UI5Type;
  name: string;
  collection?: boolean;
  ui5Aggregation?: UI5Aggregation;
  forHover?: boolean;
  reference?: string;
}): PropertyType[] => {
  /* istanbul ignore next */
  const {
    collection = false,
    ui5Aggregation,
    forHover = false,
    context,
    type,
    name,
    reference,
  } = param;
  const aggregation = !!ui5Aggregation;
  const propertyType: PropertyType[] = [];
  switch (type.kind) {
    case "UI5Any": {
      propertyType.push({
        kind: TypeKind[type.name],
        dependents: getFromMap(dependents, name, aggregation),
        notAllowedElements: getFromMap(notAllowedElements, name, aggregation),
        collection,
        reference,
      });
      break;
    }
    case "PrimitiveType":
      propertyType.push({
        kind: TypeKind[type.name],
        dependents: getFromMap(dependents, name, aggregation),
        notAllowedElements: getFromMap(notAllowedElements, name, aggregation),
        possibleValue: {
          fixed: !!defaultBoolean.get(type.name),
          values: getFromMap(defaultBoolean, type.name, aggregation),
        },
        collection,
        reference,
      });
      break;
    case "UI5Enum":
      propertyType.push({
        kind: TypeKind.string,
        dependents: getFromMap(dependents, name, aggregation),
        notAllowedElements: getFromMap(notAllowedElements, name, aggregation),
        possibleValue: {
          fixed: true,
          values: type.fields.map((field) => {
            /**
             * filter operator accepts `BT` as value in XML instead of FQN `sap.ui.model.FilterOperator.BT`
             */
            if (type.name === FILTER_OPERATOR) {
              return field.name;
            }
            return ui5NodeToFQN(field);
          }),
        },
        collection,
        reference,
      });
      break;
    case "UI5Class":
      propertyType.push({
        kind: getClassKind(type.name),
        dependents: getFromMap(dependents, name, aggregation),
        notAllowedElements: getFromMap(notAllowedElements, name, aggregation),
        possibleElements: reference
          ? []
          : getPossibleElement({ context, ui5Aggregation, forHover, type }),
        possibleValue: {
          fixed: false,
          values: getPossibleValuesForClass(context, type),
        },
        collection,
        reference,
      });
      break;
    case "UI5Typedef":
      if (TypeKind[type.name]) {
        propertyType.push({
          kind: TypeKind[type.name],
          dependents: getFromMap(dependents, name, aggregation),
          notAllowedElements: getFromMap(notAllowedElements, name, aggregation),
          collection,
          reference,
        });
      }
      break;
    case "UnionType":
      for (const unionType of type.types) {
        if (unionType.kind === "ArrayType" && unionType.type) {
          propertyType.push(
            ...buildType({
              context,
              type: unionType.type,
              name,
              collection: true,
              ui5Aggregation,
              forHover,
              reference,
            })
          );
        } else {
          propertyType.push(
            ...buildType({
              context,
              type: unionType,
              name,
              collection,
              ui5Aggregation,
              forHover,
              reference,
            })
          );
        }
      }
      break;
    case "ArrayType":
      if (!type.type) {
        break;
      }
      propertyType.push(
        ...buildType({
          context,
          type: type.type,
          name,
          collection: true,
          ui5Aggregation,
          forHover,
          reference,
        })
      );
      break;
  }
  return propertyType;
};

/**
 * Retrieves the constructor parameter properties.
 *
 * @param {Object} param - The parameter object.
 * @param {BindContext} param.context - The binding context.
 * @param {UI5ConstructorParameters[]} param.parameterProperties - The UI5 constructor parameters.
 * @param {UI5Aggregation} [param.ui5Aggregation] - The UI5 aggregation.
 * @param {boolean} [param.forHover=false] - Indicates if the properties are for hover.
 * @param {UI5Class} param.type - The UI5 class type.
 * @returns {BindingInfoElement[]} The array of binding information elements.
 */
export function getConstructorParameterProperties(param: {
  context: BindContext;
  parameterProperties: UI5ConstructorParameters[];
  ui5Aggregation?: UI5Aggregation;
  forHover?: boolean;
  type: UI5Class;
}): BindingInfoElement[] {
  const result: BindingInfoElement[] = [];
  const {
    ui5Aggregation,
    forHover = false,
    type,
    context,
    parameterProperties,
  } = param;
  for (const param of parameterProperties) {
    if (!param.type) {
      continue;
    }
    // add reference to type and avoid recursion
    const paramType = param.type;
    const reference = getReference(paramType);
    const data: BindingInfoElement = {
      name: param.name,
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      type: buildType({
        context,
        type: paramType,
        name: param.name,
        collection: false,
        ui5Aggregation,
        forHover,
        reference,
      }),
      documentation: getDocumentation({
        context,
        prop: param,
        FQN: ui5NodeToFQN(type),
        titlePrefix: "(class)",
        forHover,
      }),
    };
    data.required = !param.optional;
    result.push(data);
  }

  return result;
}
