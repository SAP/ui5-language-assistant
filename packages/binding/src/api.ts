import type { BindingIssue } from "./types";
import { BINDING_ISSUE_TYPE } from "./types";

export { getCompletionItems } from "./services/completion";
export { bindingValidators } from "./services/diagnostics";
export { getHover } from "./services/hover";
export type { BindingIssue } from "./types";

export function isBindingIssue<T extends { issueType: string }>(
  issue: BindingIssue | T
): issue is BindingIssue {
  return issue.issueType === BINDING_ISSUE_TYPE;
}

export {
  AGGREGATION_BINDING_INFO,
  PROPERTY_BINDING_INFO,
  FILTER_OPERATOR,
} from "./constant";

export { getBindingElements } from "./definition/definition";
