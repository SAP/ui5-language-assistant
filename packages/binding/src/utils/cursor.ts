import {
  BindingParserTypes as BindingTypes,
  positionContained,
  isAfterAdjacentRange,
} from "@ui5-language-assistant/binding-parser";
import { CursorContext } from "../types";
import { TextDocumentPositionParams } from "vscode-languageserver-protocol";

export const getCursorContext = (
  parm: TextDocumentPositionParams,
  binding: BindingTypes.StructureValue,
  spaces: BindingTypes.WhiteSpaces[]
): CursorContext => {
  /* istanbul ignore next */
  const { elements } = binding;
  if (elements.length === 0) {
    return {
      type: "empty",
      kind: "properties",
    };
  }
  const { position } = parm;
  // search all element
  const el = elements.find((item) => positionContained(item.range, position));
  if (el) {
    // check key
    if (el.key && positionContained(el.key.range, position)) {
      return {
        type: "key",
        kind: "properties-excluding-duplicate",
        element: el,
      };
    }
    // check colon => value
    if (el.colon && positionContained(el.colon.range, position)) {
      return {
        type: "value",
        kind: "value",
        element: el,
      };
    }
    // check value
    if (el.value && positionContained(el.value.range, position)) {
      return {
        type: "value",
        kind: "value",
        element: el,
      };
    }
  }
  // search white spaces
  const spaceEl = spaces.find((item) =>
    positionContained(item.range, position)
  );
  if (spaceEl) {
    // further check parts of adjacent element
    for (const el of elements) {
      // after adjacent key => value
      if (el.key && isAfterAdjacentRange(spaceEl.range, el.key.range)) {
        // this happen when colon is missing
        return {
          type: "value",
          kind: "value",
          element: el,
        };
      }
      // after adjacent colon => value
      if (el.colon && isAfterAdjacentRange(spaceEl.range, el.colon.range)) {
        return {
          type: "value",
          kind: "value",
          element: el,
        };
      }
    }
  }
  return {
    type: "key-value",
    kind: "properties-with-value-excluding-duplicate",
  };
};
