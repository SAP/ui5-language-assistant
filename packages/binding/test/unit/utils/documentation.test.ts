import { Context } from "@ui5-language-assistant/context";
import { getDocumentation } from "./../../../src/utils";
import { UI5TypedefProp } from "@ui5-language-assistant/semantic-model-types";
import {
  buildUI5Typedef,
  buildUI5Namespace,
} from "@ui5-language-assistant/test-utils";

const testNS = buildUI5Namespace({ name: "test" });

const testInnerNS = buildUI5Namespace({
  name: "inner",
  parent: testNS,
});

describe("documentation", () => {
  it("check PrimitiveType", () => {
    const prop = buildUI5Typedef({
      name: "TestTypedef",
      type: {
        kind: "PrimitiveType",
        name: "String",
      },
    }) as unknown as UI5TypedefProp;

    const result = getDocumentation(
      { ui5Model: {} } as unknown as Context,
      prop
    );
    expect(result).toMatchSnapshot();
  });
  it("check UI5Typedef, UI5Class or UI5Enum", () => {
    const prop = buildUI5Typedef({
      name: "TestTypedef",
      type: {
        kind: "UI5Typedef",
        name: "PropertyBindingInfo",
        parent: testInnerNS,
      },
    }) as unknown as UI5TypedefProp;
    const result = getDocumentation(
      { ui5Model: {} } as unknown as Context,
      prop
    );
    expect(result).toMatchSnapshot();
  });
  it("check UnionType", () => {
    const prop = buildUI5Typedef({
      name: "TestTypedef",
      type: {
        kind: "UnionType",
        types: [
          {
            kind: "UI5Typedef",
            name: "PropertyBindingInfo",
            parent: testInnerNS,
          },
        ],
      },
    }) as unknown as UI5TypedefProp;
    const result = getDocumentation(
      { ui5Model: {} } as unknown as Context,
      prop
    );
    expect(result).toMatchSnapshot();
  });
  it("check ArrayType", () => {
    const prop = buildUI5Typedef({
      name: "TestTypedef",
      type: {
        kind: "ArrayType",
        type: {
          kind: "UI5Typedef",
          name: "PropertyBindingInfo",
          parent: testInnerNS,
        },
      },
    }) as unknown as UI5TypedefProp;
    const result = getDocumentation(
      { ui5Model: {} } as unknown as Context,
      prop
    );
    expect(result).toMatchSnapshot();
  });
  it("check undefined type", () => {
    const prop = buildUI5Typedef({
      name: "TestTypedef",
    }) as unknown as UI5TypedefProp;

    const result = getDocumentation(
      { ui5Model: {} } as unknown as Context,
      prop
    );
    expect(result).toMatchSnapshot();
  });
});
