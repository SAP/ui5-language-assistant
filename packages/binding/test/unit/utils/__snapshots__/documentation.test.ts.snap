// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentation check ArrayType 1`] = `
Object {
  "kind": "markdown",
  "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** test.inner.PropertyBindingInfo

**Description:** undefined

**Optional:** undefined

[More information](https://ui5.sap.com/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
}
`;

exports[`documentation check PrimitiveType 1`] = `
Object {
  "kind": "markdown",
  "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** String

**Description:** undefined

**Optional:** undefined

[More information](https://ui5.sap.com/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
}
`;

exports[`documentation check UI5Typedef, UI5Class or UI5Enum 1`] = `
Object {
  "kind": "markdown",
  "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** test.inner.PropertyBindingInfo

**Description:** undefined

**Optional:** undefined

[More information](https://ui5.sap.com/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
}
`;

exports[`documentation check UnionType 1`] = `
Object {
  "kind": "markdown",
  "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** test.inner.PropertyBindingInfo

**Description:** undefined

**Optional:** undefined

[More information](https://ui5.sap.com/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
}
`;

exports[`documentation check undefined type 1`] = `
Object {
  "kind": "markdown",
  "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** 

**Description:** undefined

**Optional:** undefined

[More information](https://ui5.sap.com/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
}
`;
